Signature cleanup

ShellyHerself opened the pull request:
You will see a high number of files edited. But that's mostly one or two line changes.
There is also a high number of commits, but that is because of rapid prototyping and me wanting to keep code from good states safe.
 - Update build script to support building C++ files out of yaml files.
   ~~.gitignore now gets autoupdated with all the yaml generated files.~~
 - Created a new LiteSignature class that is more sane than CodeSignature
 - New CodeGen generates C++ signature objects from yaml.
 - Added big yaml file with _most_ of the signatures.
It's not all of them yet sadly.
I'll narrow them all down by the next pullrequest when I actually switch them over.
 - Removed file_handle_leak fix. (I don't think it actually does anything).
Sorry about that one. I thought I was going to switch all the code that uses signatures to this new code. But I decided that I'd rather have this pullrequest focus on the important stuff.

The Perl may seem like a lot. But once you get one subroutine, you'll quickly get them all.

ShellyHerself committed:
[codegen] Initial commit (Safety commit)
Contains the first draft of code generation procedures. Just prints to 
the terminal right now and is larely a test.

ShellyHerself committed:
[memory] Initial signatures yaml list
Not sure if this is all of them yet.
Might change the formatting to make supporting multiple codebases 
easier.

ShellyHerself committed:
Add copyright headers

ShellyHerself committed:
[memory] Allign comments better

ShellyHerself committed:
[fixes] Remove file descriptor leak fix as this aint it chief.
Never trust that what others did actually works. Do extensive tests.

ShellyHerself committed:
[memory] Add comment to signatures.yaml

ShellyHerself committed:
[util] Added some spacing at the end of the cutie

ShellyHerself committed:
[hooker] Add LiteSignature, future replacement for CodeSignature
This is a really cut down version of CodeSignature with the only purpose 
being to find code rather than store a ton of debug information.
This is happening because we're putting all our signatures into one file 
to get them out of source files.

ShellyHerself committed:
[codegen] Move text helpers into a module and start on signature code

ShellyHerself committed:
[memory] Specify multi signatures

ShellyHerself committed:
[codegen] split signature validation into its own string

ShellyHerself committed:
[codegen] Implement getters

ShellyHerself committed:
[memory] Add some signature properties

ShellyHerself committed:
[codegen] Make generated code nicer spaced

ShellyHerself committed:
[codegen] Restructure

ShellyHerself committed:
[codegen] Split signature logic and move it into its own module

ShellyHerself committed:
[codegen] Be more strict

ShellyHerself committed:
[codegen] Refactor signature file generation

ShellyHerself committed:
[codegen] Remove some redundant calls

ShellyHerself committed:
[build] Set build script up for yaml generation
build.pl will now direct generate.pl to build cpp sources and headers. 
It also updates the gitignore for these files.

ShellyHerself committed:
[codegen] Finalize signature handling

ShellyHerself committed:
[codegen] Clean up for pullrequest

ShellyHerself committed:
[codegen] More misc cleanup and small bug fixes

ShellyHerself committed:
[build] Style

ShellyHerself committed:
[codegen] Pretty output

ShellyHerself committed:
[hooker] Fix LiteSignature::search_multiple not having a return value
Also fix some input types

ShellyHerself committed:
[codegen] Fix minor printf bug

Mimickal commented on codegen/CodeGen/Signature.pm:
```diff
@@ -0,0 +1,170 @@
+        "Since there is crucial signatures that we cannot find Vulpes "
+        "cannot continue loading.\\n"
+        "Please submit your error log to the Github issues page.\\n");
+        exit(0);
```
You're exiting with an error here. The exit code should be non-zero to tell the shell that something went wrong. This will be useful if you ever do an automated build system.

Mimickal commented on codegen/CodeGen/Signature.pm:
```diff
@@ -0,0 +1,170 @@
+    $_->{uc_name} = $_->{uc_name} || uc $_->{name};
+    $_->{offset}  = $_->{offset}  || 0;
+    $_->{type}    = $_->{type}    || "uintptr_t";
+    $_->{multi}   = $_->{multi}   || 0;
```
Perl has an `||=` operator

Mimickal commented on codegen/CodeGen/Signature.pm:
```diff
@@ -0,0 +1,170 @@
+    # crucial can be
+    if (not exists $_->{crucial}) {
+        $_->{crucial} = 0;
+    }
```
Perl also has `//=` (if undefined, set to x), so this can be written `$_->{crucial} //= 0;`

Mimickal commented on build.pl:
```diff
@@ -50,6 +51,35 @@ sub build_luajit {
+    $start_pos = index $gitignore, "\n", $start_pos;
+    # Print new gitignore with the expected cpp and hpp files at the end
+    open(GITIGNORE, ">.gitignore");
+    print GITIGNORE (substr ($gitignore, 0, $start_pos)),"\n", (join "\n", @pp_files), "\n";
```
Why do we not have a specific build directory, then add `build/*.hpp` to the `.gitignore` file? This seems needlessly complicated. Also, modifying the .gitignore file will cause the current working directory to be marked as modified in git.

ShellyHerself replied:
What do you think about the a sub directory (let's say 'codegen' or 'generated') wherever the yaml resides. And then making some dummy files that include the ones from the generated directory?
I'd like to keep the general structure in my includes the same

ShellyHerself committed:
[codegen] Change some stuff for the clarity of outputted code.
Make all signature getters have sig_ prefix. And give all signature 
objects signature_ prefix

Mimickal commented on hooker/hooker.cpp:
```diff
@@ -73,6 +73,74 @@ uintptr_t CodeSignature::address(bool recalculate) {
+        bool mismatch = false;
+        // For each address we go through our set of bytes until
+        // we get a mismatch or we reach the end of our signature.
+        for (int j=0; j < size; j++) {
```
You can remove the `break` on line 101 by making this conditional `j < size || mismatch`

Mimickal commented on hooker/hooker.cpp:
```diff
@@ -73,6 +73,74 @@ uintptr_t CodeSignature::address(bool recalculate) {
+    uintptr_t result = UINTPTR_MAX;
+
+    uintptr_t current_address = start_address;
+    while (result == UINTPTR_MAX && current_address - size <= end_address) {
```
Where is `size` declared?

Mimickal commented on hooker/hooker.cpp:
```diff
@@ -73,6 +73,74 @@ uintptr_t CodeSignature::address(bool recalculate) {
+            // element, start the next iteration.
+            if (bytes[j] == -1 || reinterpret_cast<uint8_t*>(current_address)[j] == bytes[j]) {
+                continue;
+            }
```
This loop can be simplified by inverting this logic.
```
if (bytes[j] != -1 && reinterpret_cast<uint8_t*>(current_address)[j] != bytes[j]) {
    mismatch = true;
}

Mimickal commented on hooker/hooker.cpp:
```diff
@@ -73,6 +73,74 @@ uintptr_t CodeSignature::address(bool recalculate) {
+    printf("Search for sig %s\nAddress range: %8X - %8X\n", name, start_address, end_address);
+
+    // TODO: Use constant.
+    uintptr_t result = UINTPTR_MAX;
```
Why do we use the max pointer value instead of `null` here? That would save us from needing to check `ptr != UINTPTR_MAX` outside this function (thus preventing us from leaking this implementation detail). That would make it a lot more clear what's going on outside of this function too.

Mimickal commented on codegen/CodeGen/Signature.pm:
```diff
@@ -0,0 +1,170 @@
+
+sub yaml_sig_to_cpp_sig {
+    # Convert string to individual parts and then convert them into C++ format.
+    my @bytes = map {$_ eq "??" ? "-1" : "0x$_"} split /\s+/, $_->{bytes};
```
What happens if `$_` isn't `??` or something within `00` to `FF`?

Mimickal commented on codegen/CodeGen/Signature.pm:
```diff
@@ -0,0 +1,170 @@
+}
+
+sub yaml_sig_to_cpp_address_var {
+    if ($_->{multi} || 0) {
```
`|| 0` is redundant

Mimickal commented on codegen/CodeGen/Signature.pm:
```diff
@@ -0,0 +1,170 @@
+
+sub yaml_signatures_to_cpp_definitions {
+    my $name = shift;
+    my $sigs = shift;
```
A cleaner way to write this is `my ($name, $sigs) = @_;`. This creates a sort-of pseudo-argument signature for the function.

Mimickal commented on codegen/CodeGen/Signature.pm:
```diff
@@ -0,0 +1,170 @@
+$validation_code
+
+    // Report to the user that we are missing signatures if we are.
+    if (crucial_missing.size()) {
```
`std::vector` has a function called `empty` that's more idiomatic for what you're doing here.

ShellyHerself replied:
For the empties I'll use empty. but when checking if something is not empty using size seems more appropriate.

Mimickal commented on codegen/generate.pl:
```diff
@@ -0,0 +1,72 @@
+    # basename with yaml extension
+    my $yaml_basename = basename($filename);
+
+    my $license_header = qq{/*
```
You should define this as a subroutine (bootleg constant) at the top of this script instead.

Mimickal commented on hooker/hooker.cpp:
```diff
@@ -73,6 +73,74 @@ uintptr_t CodeSignature::address(bool recalculate) {
+
+    std::vector<uintptr_t> addresses;
+    // TODO: Use constant.
+    uintptr_t last_result = 0;
```
Is `null` not a formal thing in C++?

ShellyHerself committed:
[hooker] return NULL when signature address cannot be found

ShellyHerself committed:
[codegen] Make header a sub routine

ShellyHerself committed:
[codegen] Generate files in subfolder

ShellyHerself committed:
[codegen] Validate bytes in signatures

ShellyHerself committed:
[codegen] Add some missing spacing

ShellyHerself committed:
[meta] Add gitignore comment

ShellyHerself committed:
[memory] Add comments stating that contents of some files are autogenned

ShellyHerself committed:
[memory] Fix comments

ShellyHerself committed:
[codegen] Fix regex
It wasn't broken, but it wasn't as flexible as I wanted it. Now I've 
gotten some good sleep I finally realized how to fix it

gbMichelle commented:
So, I think this is good at this point. I was thinking of just merging it. But I don't want to overstep you in that.
